### PR TITLE
Implement copy from external table

### PR DESCRIPTION
**Pull request by acquamarin**
# Description

This PR refactors the `DuckDBScanSharedState` to inherit from `BaseScanSharedState`. So we can do copy from a external duckdb, sqlite, postgres table.

### CODE REVIEW

File: extension/duckdb/test/test_files/duckdb.test [hunk 1]
```diff
@@ -12,6 +12,21 @@ Binder exception: Unsupported duckdb type: ENUM('sad', 'ok', 'happy').
 -STATEMENT ATTACH '${KUZU_ROOT_DIRECTORY}/extension/duckdb/test/duckdb_database/tinysnb.db' as tinysnb (dbtype duckdb, skip_unsupported_table = true);
 ---- 1
 Attached database successfully.
+-STATEMENT create node table person_copy (ID INt64, fName StRING, gender INT64, isStudent BoOLEAN, isWorker BOOLEAN, age INT64, eyeSight DOUBLE, birthdate DATE, registerTime TIMESTAMP, lastJobDuration interval, workedHours INT64[], usedNames STRING[], courseScoresPerTerm INT64[][], height float, u UUID, grades INT64[4], PRIMARY KEY (ID));
+---- ok
+-STATEMENT COPY person_copy FROM tinysnb.person;
```
Comment (andyfengHKU) [inline extension/duckdb/test/test_files/duckdb.test:17]: Let's also add a test for COPY rel table maybe.

File: tools/python_api/test/test_scan_pandas.py [hunk 2]
```diff
@@ -378,6 +378,7 @@ def test_scan_long_utf8_string(tmp_path: Path) -> None:
     result = conn.execute("LOAD FROM df WHERE name = '非常长的中文' RETURN count(*);")
     assert result.get_next() == [1]
 
+
```
Comment (andyfengHKU) [inline tools/python_api/test/test_scan_pandas.py:381]: remove new line

File: src/include/function/table/scan_functions.h [hunk 3]
```diff
@@ -18,30 +18,31 @@ struct BaseScanSharedState : public TableFuncSharedState {
...
 
-    explicit BaseFileScanSharedState(uint64_t numRows)
-        : BaseScanSharedState{}, fileIdx{0}, blockIdx{0}, numRows{numRows} {}
+    explicit BaseScanSharedStateWithNumRows(uint64_t numRows) : numRows{numRows} {}
 
     uint64_t getNumRows() const override { return numRows; }
 };
 
-struct ScanSharedState : public BaseFileScanSharedState {
+struct ScanSharedState : public BaseScanSharedStateWithNumRows {
```
Comment (andyfengHKU) [inline src/include/function/table/scan_functions.h:29]: Let me add my thoughts here. I think it make sense to have

```
TableScanSharedState {
// empty
}

// Used for scan function that scans a single tabular object
TableScanSharedStateWithRowInfo : TableScanSharedState {
lock
numRows
}

// Used when we scan multiple files
TableScanSharedStateWithFileInfo : public TableScanSharedStateWithRowInfo {
readerConfig
fileIdx
blockidx
}

Ideally we wanna merge ScanFileSharedState into TableScanSharedStateWithFileInfo and let @MSebanc solves the issue about unifying progress bar printing.
```

**Review by andyfengHKU — APPROVED** [review]

File: tools/python_api/test/test_scan_pandas.py (hunk 2 above)
Comment (acquamarin) [reply tools/python_api/test/test_scan_pandas.py:381]: i think the convension in python is two newlines between two functions.

**Pull request merged**